### ISSUE TITLE
upx.exe 3.96 quarantined / compression issues

### ISSUE THREAD

**Issue by covid19era** [upx/upx#359]
## What's the problem (or question)?

Malwarebytes detected a threat from packed upx.exe and would moved it to quarantine by default. 

unpacking upx.exe solved the problem.

but most importanly, unpacking compressed files have benefits of... compressing multiple executable files into one "solid" archive, which is impossible if each executable is compressed by UPX. many software comes with multiple executables and libraries.

tip: Windows 10 have its own packer for data and executable files: "compact" program, quote: "supported algorithms are: XPRESS4K (fastest) (default) XPRESS8K XPRESS16K LZX (most compact)" - not as strong as LZMA, but leaves files intact and ready for use. (requires NTFS formatted with 4KB cluster size.)

## What should have happened?

1. make default behavior for UPX to unpack packed files intead to pack them.
2. maybe storing original file's checksum and additional recovery data inside packed file is a good idea?

## Do you have an idea for a solution?

also, use option --strip-relocs=0 when unpacking by default, as unpacked execs may not work otherwise.

## How can we reproduce the issue?

1. i made this test aside of all of the above;
2. compress 7z.exe from 20.00 alpha package with UPX 3.96 (default or --lzma);
3. upcompress 7z.exe without "--strip-relocs=0" parameter;
4. uncompressed 7z.exe doen't work!

## Please tell us details about your environment.

* UPX version used (`upx --version`): 3.96
* Host Operating System and version: windows 10 pro x64
* Host CPU architecture: amd64-win64.pe
* Target Operating System and version: windows 10 pro x64
* Target CPU architecture: amd64-win64.pe, i386-win32.pe, amd64-linux.elf
[7z.zip](https://github.com/upx/upx/files/4464151/7z.zip)

**Comment by jreiser** [comment]
The quarantine issue is same as https://github.com/upx/upx/issues/337 .  That anti-virus vendor (Malwarebytes) is behind the times, they do not recognize upx-3.96.  Contact Malwarebytes.

As for "What should have happened?", UPX applies the same processing outline to each PE file, and will pay no attention to processing by any prior "compact" program.  Also, UPX already stores the adler32 checksum of the original file.
As for "How can we reproduce the issue", please upload ("Attach files by ..."), or supply explicit literal URLs, for input files; UPX developers will not go searching for them.  Also, if you want UPX to decompress then you must specify `-d` or `--decompress`.

**Comment by covid19era** [comment]
@jester
your reply makes no sence.
1. anti-virus software uses machine learning, heuristics and such. first thing they learn is that >99% of malware is exe-packed. no point contacting Malwarebytes with compressed exe.
2. windows "compact" program meant to be transparent while reading data from file normally. (btw, LZX compression ratio is similar to ZIP.)
3. if UPX stores checksum, why it does not use it? unpacked execs are damaged and UPX didn't even warn about.
4. i'm sure UPX developers have enough program samples which can't be unpacked without "--strip-relocs=0" parameter already.
5. i know how to use "-d".

**Comment by jreiser** [comment]
1. Which version/edition/malware_definitions was in effect when Malwarebytes quarantined upx-3.96.exe?  New "virus signatures" (positive and negative) appear frequently.  Comments on Issue #337 show that over a period of a few weeks several anti-malware utilities changed their mind about upx-3.96.exe, as reported by VirusTotal. Just because an executable is exe-packed does NOT imply it is malware; for example, the result of applying UPX to any non-malware program.  All pre-compiled versions of upx.exe that are provided by UPX Team have been packed by UPX, but none are malware.  The better anti-malware utilities have a channel for "appeal" of their classification.  One of the criteria for reversing a decision can be the number of reasonable appeals; so another "vouching" for upx-3.96.exe (include its SHA256 with your appeal) might convince Malwarebytes to change its verdict sooner.
2. Stating a couple good URLs for the "windows compact program" would make communication more effective and efficient.  What is its "value proposition"?  How is it used?  Is it much the same as gzopen() from zlib (https://www.zlib.net/manual.html)?  If so, then it is NOT 100% transparent: a program which reads the file must know to look for applicability of gzopen.  Such code is heuristic, and suffers from "false positive" errors much like Malwarebytes.  Furthermore, the default composition of "windows compact program" and UPX is readily achievable by a simple shell script. (If not then "windows compact program" omits a vital feature: un-compact.)
3. See next paragraph 4).
4. The "axiom of choice" ("There are so many examples; just pick one") does NOT apply to debugging.   The issue creator must supply a specific literal actual example.  That's the ONLY way to be sure that everyone is dealing with the same situation.  It is likely that there are dozens of "free variables" that the issue creator fails to specify in the Description of the issue.  A specific literal example, together with an observation of the reproducible actual behavior versus the expected behavior, is by far the most effective way to focus attention.  If that is not worth the time of the issue creator, then it is quite likely not worth the effort of the maintainer to investigate and resolve.

**Comment by covid19era** [comment]
@jreiser 

1. just don't compress UPX. otherwise you cannot guarantee that a new anti-virus database update won't find UPX "malicious" again.

2. it is not a bug, it's a "feature" because we have "--strip-relocs=0" parameter for some reason. UPX "strips relocations" when unpacking, but many files don't like it. you cannot fine-tune UPX for every new executable, just make "--strip-relocs=0" parameter default when decompressing UPX files.

3. Windows "compact" operates just like NTFS volume compression. mode of operation similar to old DIET compression "Terminate-and-Stay-Resident" program for MS-DOS. file stays "compacted" until it is modified. "compact" is pre-intalled in every new Windows i think. LZX ratio slighly better than ZIP/GZIP/ZLIB. you can see compressed file size via file properties in Windows Explorer, just like NTFS compression, but attribute "compressed" might be absent and file's time "touched".

there's nothing more to discuss...

**Comment by levicki** [comment]
> anti-virus software uses machine learning, heuristics and such. first thing they learn is that >99% of malware is exe-packed. no point contacting Malwarebytes with compressed exe.

Antivirus software is capable of unpacking UPX executables without executing them (just like UPX itself does) and then scanning the unpacked file for malicious code to avoid false positives. At least that's what good antivirus software does.

> just don't compress UPX. otherwise you cannot guarantee that a new anti-virus database update won't find UPX "malicious" again.

How about you don't use antivirus programs that report false positives, and use something more reliable instead? Like ESET, or even Windows Defender.

> ... because we have "--strip-relocs=0" parameter for some reason. UPX "strips relocations" when unpacking...

The option `--strip-relocs` is used during packing, not unpacking. You obviously have no clue what you are talking about. Try reading documentation on how to properly use UPX.

> Windows "compact" operates just like NTFS volume compression...

Windows' `compact.exe` doesn't operate like NTFS volume compression -- it is directly controlling NTFS file and folder compression.

Compressing a program using NTFS compression does not change its internal structure -- it is an equivalent of packing an EXE file into a ZIP archive.

Copying NTFS compressed EXE to USB drive or uploading it to a web site, or sending it via email decompresses it before the operation and what the other side gets is decompressed.

In other words, not even remotely similar to what UPX does, not to mention that it is not useful at all for software distribution and requires NTFS to work which doesn't exist on Linux or Mac.

> there's nothing more to discuss...

Indeed there isn't:

1. You don't understand the PE executable format enough to understand why packing some executables might need different options (and why it might not even work for others).
2. You don't respect developers (you did not follow the issue submission rules).
3. You are rude and confrontational, and your chosen username is bordering between bad taste and offensive.

**Comment by bitraid** [comment]
>The option --strip-relocs is used during packing, not unpacking

It *is* used for unpacking, too.  The relocation table is required for ASLR enabled executables (linked with `/DYNAMICBASE`, which sets the` IMAGE_DLLCHARACTERISTICS_DYNAMIC_BASE` header flag - the default with latest Visual Studio).
Other that that, i agree with everything else.

**Comment by levicki** [comment]
> It _is_ used for unpacking, too.

Sorry if that bit wasn't clear, What I meant is that if relocs are stripped during packing (which is the default behavior according to documentation), then providing that option during unpacking can't possibly mean anything unless UPX is actually able to reconstruct relocs which were stripped.

**Comment by bitraid** [comment]
I think there are currently some issues with relocation stripping and  it doesn't work as it should.
The relocation section is not actually removed - unpacking a file with `--strip-relocs=0`, will result in a file with its original relocation section, even if it was packed with `--strip-relocs=1`. If unpacked with `--strip-relocs=1`, the unpacked file still has a relocation section (size not `0`) but with no data.

**Comment by levicki** [comment]
@bitraid I think that this behavior is confusing. Here is what I think should happen:

1. If an image with ASLR must have relocations, then `--strip-relocs` option **should do nothing both when packing and unpacking**.
2. If an image is packed with `--strip-relocs=1`, then it should not have reloc section when unpacked.
3. If an image is packed with `--strip-relocs=0`, then it should not be removed if you specify `--strip-relocs=1` during unpacking.

In short, I think the best way to accomplish this is to make `--strip-relocs` to be only packing option, and to do nothing for ASLR enabled images but emit a warning.

**Comment by bitraid** [comment]
@levicki That makes sense. Should the warning be just for ASLRs though, or for every PE where the relocations cannot be removed (which would include DLLs and 64-bit executables)? That would ofcourse result to a lot of warnings. Instead, maybe just add a "where possible" at the documentation of `--strip-relocs=1`.

**Comment by covid19era** [comment]
attached file to the first post for testing if you are curious.

(file needs to be compressed and then decompressed using both methods.)

btw, there's "win32 relocations rebuilder v1.0 by NEOx" for UPX-decompressed files written in 2003, but it worked today with the UPX 3.96 problem. (after rebuild, file became bigger than original)

**Comment by bitraid** [comment]
> attached file to the first post for testing if you are curious.
> 
> (file needs to be compressed and then decompressed using both methods.)
> 
> btw, there's "win32 relocations rebuilder v1.0 by NEOx" for UPX-decompressed files written in 2003, but it worked today with the UPX 3.96 problem. (after rebuild, file became bigger than original)

@covid19era Still, you don't seem understand what's happening here. In short: `--strip-relocs=1` is ignored when packing ASLR executables (since relocations are required) but not ignored for unpacking (so the unpacked program won't run, unless `--strip-relocs=0` is set). This happens for all ASLR executables, even for upx itself, when built with msvc-14 (where linked with `/DYNAMICBASE`): https://github.com/upx/upx-automatic-builds/tree/devel-20200215-b72eac6-appveyor

**Comment by levicki** [comment]
> Should the warning be just for ASLRs though, or for every PE where the relocations cannot be removed (which would include DLLs and 64-bit executables)?

I'd say the latter.

> That would ofcourse result to a lot of warnings.

Or given that today there are more executables that depend on relocations to work correctly, perhaps the best option would be to change the default behavior of UPX to not strip relocs?

Then you would have to issue warning only when user explicitly uses the option and the executable requires relocations to work. Something to the tune of:

```
EXAMPLE.DLL is ASLR enabled, ignoring --strip-relocs=1
```

So to summarize:

1. Make `--strip-relocs` a packing only option.
2. Make `--strip-relocs` default to `0` instead of `1`.
3. Make `--strip-relocs` actually fully remove the `.reloc` section when possible.
4. Print a "switch ignored" warning when `--strip-relocs=1` is passed during packing and current executable requires relocs to work.
5. Clarify when and why it is ignored in the documentation.

In my opinion that would vastly improve the current confusing behavior.

**Comment by bitraid** [comment]
>Or given that today there are more executables that depend on relocations to work correctly, perhaps the best option would be to change the default behavior of UPX to not strip relocs?

I don't think that the existance of more ASLR executables justifies the the change of not removing the relocation section of any other exe by default.

>`EXAMPLE.DLL is ASLR enabled, ignoring --strip-relocs=1`

DLLs always need relocations (ASLR or not).

>In my opinion that would vastly improve the current confusing behavior.

The confusion comes from a bug that prevents the restore of the section when unpacking aslr-exes (i opened #361 with a fix) and the lack of documentation. The complete deletion of  `.reloc`, as well as the stripping of PE32+, is not an oversight - it just needs work to be done. And even though `--strip-relocs` may not seem so useful for unpacking, i don't think it hurts either (provided that it works as it should and doesn't strip the section from files that need it). Maybe it should also keep the section for any file that wasn't stripped on compression, unless it's explicitly set by the user. Let's see if the developers have something to say about it.

**Comment by jreiser** [comment]
The design intent of off-line de-compression `upx -d` is to reconstruct the file that originally was given to UPX for compression.  The primary goal is to show that UPX should not be considered an "accomplice to malware", and to co-operate with anti-malware software.  A secondary goal is to provide an easy way to reconstruct original debugging info that is not necessary for normal execution.  Some software projects have one or more ways to separate an executable from its compiled debug information (symbol names and types, mapping from program counter to file+line number , etc.), but it can be convenient for an individual executable to keep the information together.  UPX still can save space by compressing the debuginfo because it is not needed for ordinary execution, but the debuginfo can be retrieved by `upx -d`.

Because relocation info is needed for ASLR, then UPX compression must retain `.reloc` in some form, and make it available to the run-time environment at an appropriate time.  However, for a non-ASLR module (fixed-address main program, or fixed-address .dll) then most of `.reloc` is [should be] superfluous for execution, and thus could remain compressed except for `upx -d`.

So `--strip-relocs` should have no effect on `upx -d`, except possibly to draw a warning about being ignored.  As for compression of a non-ASLR module, then ideally the software build chain  should have facilities for removing "non-essential" `.reloc` _before_ applying UPX compression.  [On Linux the static binder `/bin/ld` and utiliy programs `strip` and `objcopy` have various facilities for removing or manipulating symbol and relocation tables.]  Thus `--strip-relocs` could be interpreted during UPX compression as permission to discard unneeded information, such that `upx -d` would omit that information.

It seems to me that pull request https://github.com/upx/upx/pull/361 is a step forward towards resolving this issue.  I will apply it to the `devel` branch of UPX.

**Comment by covid19era** [comment]
it's not confusing UPX behavior, it's just executable packer behavior.
all executable packers are the same weird stuff.

UPX keeps relocs inside packed exe.
when unpacking, UPX overwrites unpacked relocs with zeros.
if you pack the exe again, it will be 1KB smaller in size (no relocs - finally stripped!)

**Comment by covid19era** [comment]
@bitraid i didn't, it was default UPX

**Comment by bitraid** [comment]
@covid19era That's the title of a linked pull request - it's not a statement directed to a person.

**Comment by bitraid** [comment]
>Because relocation info is needed for ASLR, then UPX compression must retain .reloc in some form, and make it available to the run-time environment at an appropriate time. However, for a non-ASLR module (fixed-address main program, or fixed-address .dll) then most of .reloc is [should be] superfluous for execution, and thus could remain compressed except for upx -d.

@jreiser  I must clarify here that the relocations section is always compressed (in its compressed form, the executable doesn't have a `.reloc` section). When compressing with `--strip-relocs=1` flag (default), the header flag `IMAGE_FILE_RELOCS_STRIPPED` will be set to the compressed file, and the `.reloc` section will be zeroed out (ideally it should not include the section at all). Once this is done, unpacking the file will also produce an executable with stripped relocations (although, the `.reloc` section will still be there, having it's original size but no data). The packer already ignores `--strip-relocs=1` for DLLs and ASLR-EXEs, where the section is required (currently, the parameter is ignored for PE32+, too). With #362, the unpacker always restores the section (and data) when it's not stripped, which i think is the logical thing to do.

**Comment by levicki** [comment]
@bitraid @jreiser I see the pull requests that correct reloc stripping behavior have beem merged into devel branch and tagged as v4.0.0 milestone. However, it's been two years since then and UPX is still on 3.96. Can we have a release (say 3.97) that includes those fixes? Unless 4.0 is about to release soon of course.
